counsel.el

```diff
@@ -5186,16 +5186,37 @@ N in your system."
   "List of potential build subdirectory names to check for."
   :type '(repeat directory))
 
+(defvar counsel-compile-phony-pattern
+  (rx (: bol ".PHONY:" space (group (one-or-more print)) eol))
+  "Regexp for extracting PHONY targets from Makefiles.")
+
 ;; This is loosely based on the Bash Make completion code
+(defun counsel-compile--probe-make-targets (dir)
+  "Return a list of Make targets for DIR.
+
+If the make errors we return an empty list.  This is usually because
+some sort of configuration needs to be done first or the source tree
+is pristine and being used for multiple build trees."
+  (let ((default-directory dir)
+        (targets))
+    (with-temp-buffer
+      ;; 0 = no-rebuild, -q & 1 needs rebuild, 2 error
+      (when (> 2 (call-process "make" nil t nil "-nqp"))
+        (goto-char (point-min))
+        (while (re-search-forward counsel-compile-phony-pattern nil t)
+          (setq targets
+                (nconc targets
+                       (split-string (match-string-no-properties 1)))))))
```

Comment:
Isn't it faster to `push` the split strings in a loop and `nreverse` & `nconc` the result at the end?

Reply:
I guess... I'll tweak it. TBH there seems to be be so many ways of manipulating lists I get a bit lost which is why I tend to stick to a relatively consistent abstraction like dash in my personal code. But then again it's never usually much of a performance issue ;-)

Reply:
The many ways of manipulating lists aside, perhaps it will help to remember that one of the most fundamental idioms when dealing with lists (especially in Lisp and Haskell) is to minimise how many times they're traversed. In this example, the `nconc` in a loop repeatedly traverses the ever-growing `targets` list.

Reply:
Hmm I can't seem to push from split-string directly:

    (defun counsel-compile--probe-make-targets (dir)
      "Return a list of Make targets for DIR.

    Return an empty list is Make exits with an error.  This might
    happen because some sort of configuration needs to be done first
    or the source tree is pristine and being used for multiple build
    trees."
      (let ((default-directory dir)
            (targets))
        (with-temp-buffer
          ;; 0 = no-rebuild, -q & 1 needs rebuild, 2 error (for GNUMake at
          ;; least)
          (when (> 2 (call-process "make" nil t nil "-nqp"))
            (goto-char (point-min))
            (while (re-search-forward counsel-compile-phony-pattern nil t)
              (push targets
                    (split-string (match-string-no-properties 1))))))
        (sort targets 'string-lessp)))

I get a confusing error message:

    let*: Symbol’s function definition is void: \(setf\ split-string\)

Reply:
huh I'm an idiot

Reply:
OK I'm struggling with the form you want here. I I use push I end up with lists in lists which is confusing...

Reply:
OK so I just assume targets is a list of lists....

    ;; This is loosely based on the Bash Make completion code
    (defun counsel-compile--probe-make-targets (dir)
      "Return a list of Make targets for DIR.

    Return an empty list is Make exits with an error.  This might
    happen because some sort of configuration needs to be done first
    or the source tree is pristine and being used for multiple build
    trees."
      (let ((default-directory dir)
            (targets))
        (with-temp-buffer
          ;; 0 = no-rebuild, -q & 1 needs rebuild, 2 error (for GNUMake at
          ;; least)
          (when (> 2 (call-process "make" nil t nil "-nqp"))
            (goto-char (point-min))
            (while (re-search-forward counsel-compile-phony-pattern nil t)
              (push (split-string (match-string-no-properties 1))
                    targets))))
        (sort (apply #'nconc targets) 'string-lessp)))

Reply:
1. `targets` starts as the empty list `nil`.
2. In each loop iteration, `split-string` generates a list of strings.
3. We `push` this list of strings onto the start of `targets`.
4. When the loop exits, `targets` will be a list of lists of strings in the opposite order from that of `make`'s output, so we `nreverse` it.
5. Now `targets` is a list of lists of strings in the correct order, so we flatten it in a single, efficient traversal by linking each nested list to its neighbour using `nconc`.
6. We now have a list of strings in the right order.

Here's the gist of it:

```el
(let (targets)
  (while ...
    (push (split-string ...) targets))
  (apply #'nconc (nreverse targets)))
```

If you want to sort `targets` lexicographically, then there's no need to call `nreverse`, as the order may be subsequently changed:

```el
(let (targets)
  (while ...
    (push (split-string ...) targets))
  (sort (apply #'nconc targets) #'string-lessp))
```

Reply:
Right I got there in the end... I don't care about the order in the makefile, I sort it all at the end once.